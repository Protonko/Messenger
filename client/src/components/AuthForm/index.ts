export const AuthForm = () => {
  return null
}
