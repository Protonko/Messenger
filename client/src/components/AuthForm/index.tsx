// types
import {TextTypes} from 'models/common/text'
import {FormTypes} from 'models/common/auth'
import {TInputTypes} from 'models/common/input'

import {FC, useEffect, useState} from 'react'
import {useFormik} from 'formik'
import * as yup from 'yup'
import {Text} from 'components/common/Text'
import {Input} from 'components/common/Input'
import {Button} from 'components/common/Button'

export interface IPropsAuthForm {
  title: string
  description?: string
  type?: FormTypes
}

interface IInputForm {
  id: number
  name: 'email' | 'password'
  placeholder: string
  type: TInputTypes
}

const INPUTS_DATA: Array<IInputForm> = [{
  id: 0,
  name: 'email',
  placeholder: 'Email',
  type: 'text',
}]

export const AuthForm: FC<IPropsAuthForm> = ({
  title,
  description,
  type = FormTypes.auth,
}) => {
  const [inputsData, setInputsData] = useState(INPUTS_DATA)
  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
    },
    validationSchema: yup.object({
      email: yup
        .string()
        .email('Incorrect email address')
        .required('Field is required'),
      password: yup
        .string()
        .min(6, 'The minimum password length is 6 characters')
        .required('Field is required'),
    }),
    onSubmit: (values: any) => {
      alert(JSON.stringify(values, null, 2))
    },
  })

  useEffect(() => {
    if (type === FormTypes.auth) {
      setInputsData([
        ...inputsData,
        {
          id: 1,
          name: 'password',
          type: 'password',
          placeholder: 'Password',
        }
      ])
    }
  }, [])

  const renderInputs = (elem: IInputForm) => {
    const {name, placeholder, type} = elem

    return (
      <li key={elem.id} className="auth-form__input">
        <Input
          value={formik.values[name]}
          onChange={formik.handleChange}
          name={name}
          type={type}
          placeholder={placeholder}
          error={formik.touched[name] && formik.errors[name]}
        />
      </li>
    )
  }

  return (
    <form className="auth-form" onSubmit={formik.handleSubmit}>
      <Text type={TextTypes.h3}>
        {title}
      </Text>

      {description && (
        <Text type={TextTypes.p} customStyles="auth-form__description">
          {description}
        </Text>
      )}

      <div className="auth-form__inputs">
        <ul className="list list--reset auth-form__inputs-list">
          {inputsData.map(renderInputs)}
        </ul>
      </div>

      <div className="auth-form__buttons">
        <ul className="list list--reset auth-form__buttons-list">
          <li className="auth-form__buttons-item">
            <Button
              text="Log In"
              type="submit"
              additionalClassName="auth-form__button"
            />
          </li>
          <li className="auth-form__buttons-item">
            <Button
              text="Create New Account"
              modifier="linking"
              additionalClassName="auth-form__button"
            />
          </li>
        </ul>
      </div>
    </form>
  )
}
