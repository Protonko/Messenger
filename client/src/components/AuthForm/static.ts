import {TInputTypes} from 'models/common/input'
import {FormTypes} from 'models/auth'

export interface IInputForm {
  id: number
  name: 'email' | 'password' | 'full_name'
  placeholder: string
  type: TInputTypes
}

export const FORM_DATA = {
  [FormTypes.auth]: {
    title: 'Sign in',
    description:
      'Note that you need an existing account to log in to this messenger. To sign up, use the button bellow.',
    buttonTitle: 'Log In',
    linkingButtonTitle: 'Create new account',
  },
  [FormTypes.register]: {
    title: 'Sign up',
    description: '',
    buttonTitle: 'Sign Up',
    linkingButtonTitle: 'I already have an account',
  },
}

export const INPUTS_DATA: Array<IInputForm> = [
  {
    id: 0,
    name: 'email',
    placeholder: 'Email',
    type: 'text',
  },
  {
    id: 1,
    name: 'password',
    type: 'password',
    placeholder: 'Password',
  }
]

export const INPUT_NAME_DATA: IInputForm = {
  id: 2,
  name: 'full_name',
  type: 'text',
  placeholder: 'Your name',
}
