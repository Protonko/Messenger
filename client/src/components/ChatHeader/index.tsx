import type {FC} from 'react'
import type {RootState} from 'store/reducers'
import {useSelector} from 'react-redux'
import {ReactComponent as Phone} from 'assets/icons/phone.svg'
import {TextTypes} from 'models/common/text'
import {Text} from 'components/common/Text'
import {useSearchParams} from 'hooks/useSearchParams'

interface IChatHeaderProps {
  onPhoneClick: () => void
}

export const ChatHeader: FC<IChatHeaderProps> = ({onPhoneClick}) => {
  const dialogParam = useSearchParams('dialog')
  const {dialogs} = useSelector((state: RootState) => state.dialogs)
  const interlocutor = dialogs?.find(({id}) => id === dialogParam)?.interlocutor

  if (interlocutor) {
    return (
      <div className="chat-header">
        <div className="chat-header__text">
          <Text type={TextTypes.h4} customStyles="chat-header__title">
            {interlocutor.full_name ?? ''}
          </Text>

          <Text type={TextTypes.p} customStyles="chat-header__description">
            {`(Last seen: ${new Date(interlocutor.last_seen).toLocaleString() ?? ''})`}
          </Text>
        </div>

        <Phone
          className="chat-header__call-icon"
          onClick={onPhoneClick}
        />
      </div>
    )
  }

  return <div className="chat-header" />
}
