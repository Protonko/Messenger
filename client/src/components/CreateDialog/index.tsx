import type {RootState} from 'store/reducers'
import type {IUser} from 'models/user'
import {useRef, useState} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {ReactComponent as PencilPaper} from 'assets/icons/pencil-paper.svg'
import {TextTypes} from 'models/common/text'
import {getUsers, resetUsersState, setSelectedUserId} from 'store/actions/users'
import {Modal} from 'components/common/Modal'
import {Text} from 'components/common/Text'
import {User} from 'components/common/User'
import {Button} from 'components/common/Button'
import {Textarea} from 'components/common/Textarea'
import {ContentContainer} from 'components/common/ContentContainer'

export const CreateDialog = () => {
  const {users, loading, errorMessage} = useSelector((state: RootState) => state.users)
  const dispatch = useDispatch()
  const [modalStepOneVisibility, setModalStepOneVisibility] = useState(false)
  const [modalStepTwoVisibility, setModalStepTwoVisibility] = useState(false)
  const buttonRef = useRef(null)
  const usersRef = useRef(null)

  const renderUsers = (user: IUser) => {
    return (
      <li className="users__item" key={user.id} onClick={() => onClickUser(user.id)}>
        <User name={user.full_name} src={user.avatar ?? ''} description={user.last_seen.toString()} />
      </li>
    )
  }

  const onClickUser = (id: string) => {
    setModalStepOneVisibility(false)
    setModalStepTwoVisibility(true)
    dispatch(setSelectedUserId(id))
  }

  const onOpenModal = () => {
    dispatch(getUsers())
    setModalStepOneVisibility(true)
  }

  return (
    <>
      <PencilPaper
        ref={buttonRef}
        className="dialogs__header-button"
        onClick={onOpenModal}
      />

      {/* Step one - select users modal */}
      <Modal
        customStyles="dialogs-modal"
        initiatorRef={buttonRef}
        modalVisibility={modalStepOneVisibility}
        toggleVisibilityModal={setModalStepOneVisibility}
        width={420}
      >
        <header className="dialogs-modal__section dialogs-modal__section--header">
          <Text type={TextTypes.h4} customStyles="dialogs-modal__title">
            Select user
          </Text>
        </header>

        <div className="dialogs-modal__section dialogs-modal__section--body">
          <ContentContainer loading={loading} errorMessage={errorMessage}>
            <ul className="users list list--reset" ref={usersRef}>
              {users?.length ? users.map(renderUsers) : <li className="users__item users__item--empty">Empty</li>}
            </ul>
          </ContentContainer>
        </div>
      </Modal>

      {/* Step two - write a message modal */}
      <Modal
        customStyles="dialogs-modal"
        initiatorRef={usersRef}
        modalVisibility={modalStepTwoVisibility}
        toggleVisibilityModal={setModalStepTwoVisibility}
        onClose={() => dispatch(resetUsersState())}
        width={420}
        height={300}
      >
        <header className="dialogs-modal__section dialogs-modal__section--header">
          <Text type={TextTypes.h4} customStyles="dialogs-modal__title">
            Write a message
          </Text>
        </header>

        <div className="dialogs-modal__section dialogs-modal__section--body">
          <div className="dialogs-modal__textarea">
            <Textarea rows={10} />
          </div>
        </div>

        <footer className="dialogs-modal__section dialogs-modal__section--footer">
          <Button text="Create dialog" />
        </footer>
      </Modal>
    </>
  )
}
