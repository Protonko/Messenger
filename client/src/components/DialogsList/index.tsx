// types
import {IDialog} from 'models/dialog'

import {FC} from 'react'
import hash from 'object-hash'
import {Dialog} from 'components/common/Dialog'

export interface IPropsDialogList {
  dialogs: Array<IDialog>
}

export const DialogsList: FC<IPropsDialogList> = ({dialogs}) => {
  const renderItem = (dialog: IDialog) => {
    return (
      <li className="dialogs__item" key={hash(dialog)}>
        <Dialog {...dialog} />
      </li>
    )
  }

  return (
    <div className="dialogs">
      <ul className="dialogs__list list list--reset">
        {dialogs.map(renderItem)}
      </ul>
    </div>
  )
}
