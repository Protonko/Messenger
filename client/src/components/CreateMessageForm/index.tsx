import type {EmojiData} from 'emoji-mart'
import type {RootState} from 'store/reducers'
import {useState, useCallback, useMemo, FormEvent} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {ReactComponent as Clip} from 'assets/icons/clip.svg'
import {ReactComponent as Microphone} from 'assets/icons/microphone.svg'
import {EVENTS_SOCKET} from 'models/common/socket'
import {socket} from 'utils/socket'
import throttle from 'utils/throttle'
import {useSearchParams} from 'hooks/useSearchParams'
import {TYPING_TIMEOUT} from 'static/constants'
import {createMessage} from 'store/actions/message'
import {commonError} from 'store/actions/error'
import {Textarea} from 'components/common/Textarea'
import {Avatar} from 'components/common/Avatar'
import {Button} from 'components/common/Button'
import {Emoji} from 'components/common/Emoji'

export const CreateMessageForm = () => {
  const dialogParam = useSearchParams('dialog')
  const {dialogs, account} = useSelector((state: RootState) => ({
    ...state.dialogs,
    ...state.auth,
  }))
  const dispatch = useDispatch()
  const [value, setValue] = useState('')
  const interlocutor = useMemo(
    () => dialogs?.find((dialog) => dialog.id === dialogParam)?.interlocutor,
    [dialogs, dialogParam],
  )

  const emitWriteMessage = useCallback(
    throttle(() => {
      interlocutor && socket.emit(EVENTS_SOCKET.TYPING_MESSAGE, interlocutor.id)
    }, TYPING_TIMEOUT),
    [dialogs],
  )

  const onChange = (value: string) => {
    emitWriteMessage()
    setValue(value)
  }

  const onSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()

    if (!interlocutor || !dialogParam) {
      return dispatch(commonError('ID not found!'))
    }

    dispatch(
      createMessage({
        text: value,
        dialogId: dialogParam,
        interlocutorId: interlocutor.id,
      }),
    )
  }

  const onSelect = (data: EmojiData) => {
    if ('native' in data) {
      setValue((prev) => `${prev} ${data.native}`)
    }
  }

  return (
    <form className="create-message-form" onSubmit={onSubmit}>
      <div className="create-message-form__row">
        <Avatar name={account?.full_name} src={account?.avatar ?? ''} />
        <Textarea
          placeholder="Write a message..."
          className="create-message-form__textarea"
          value={value}
          onChange={onChange}
        />
        <Avatar
          name={interlocutor?.full_name}
          src={interlocutor?.avatar ?? ''}
        />
      </div>

      <div className="create-message-form__row create-message-form__actions">
        <div className="create-message-form__col">
          <ul className="create-message-form__icons list list--reset">
            <li className="create-message-form__icons-item">
              <Clip className="create-message-form__icon" />
            </li>
            <li className="create-message-form__icons-item">
              <Microphone className="create-message-form__icon" />
            </li>
            <li className="create-message-form__icons-item">
              <Emoji onSelect={onSelect} />
            </li>
          </ul>
        </div>

        <div className="create-message-form__col">
          <Button type="submit" text="SEND" />
        </div>
      </div>
    </form>
  )
}
