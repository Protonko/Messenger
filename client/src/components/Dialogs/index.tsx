import type {IDialog} from 'models/dialog'
import type {RootState} from 'store/reducers'
import {useEffect, useState, FC, MouseEvent as ReactMouseEvent} from 'react'
import {useHistory} from 'react-router-dom'
import {useDispatch, useSelector} from 'react-redux'
import {useSearchParams} from 'hooks/useSearchParams'
import {useTyping} from 'hooks/useTyping'
import {changeReadStatus, getDialogs} from 'store/actions/dialogs'
import {socket} from 'utils/socket'
import {EVENTS_SOCKET} from 'models/common/socket'
import {Dialog} from 'components/common/Dialog'
import {Search} from 'components/common/Search'
import {ContentContainer} from 'components/common/ContentContainer'
import {CreateDialog} from 'components/CreateDialog'

export const Dialogs: FC = () => {
  const history = useHistory()
  const dialogParam = useSearchParams('dialog')
  const [typing, typingHandler] = useTyping()
  const [typingMessageAuthorId, seTypingMessageAuthorId] = useState<string>()
  const dispatch = useDispatch()
  const {account} = useSelector((state: RootState) => state.auth)
  const {dialogs, loading, errorMessage} = useSelector(
    (state: RootState) => state.dialogs,
  )

  useEffect(() => {
    socket.on(EVENTS_SOCKET.NEW_DIALOG, (dialog: IDialog) => {
      console.log(dialog)
    })

    socket.on(EVENTS_SOCKET.READ_MESSAGE, (dialogId: string) => {
      dispatch(changeReadStatus(dialogId))
    })

    socket.on(EVENTS_SOCKET.TYPING_MESSAGE, (authorId: string) => {
      seTypingMessageAuthorId(authorId)
      typingHandler()
    })
  }, [])

  useEffect(() => {
    dispatch(getDialogs())
  }, [])

  const selectDialog = (event: ReactMouseEvent<HTMLUListElement, MouseEvent>) => {
    const element = (event.target as HTMLUListElement).closest('.dialogs__item')

    if (!element) return

    history.push(`?dialog=${element.id}`)
  }

  const renderItem = (dialog: IDialog) => {
    return (
      <li
        className="dialogs__item"
        key={dialog.id}
        id={dialog.id}
      >
        <Dialog
          {...dialog}
          isTyping={typingMessageAuthorId === dialog.interlocutor.id && typing}
          isOwnMessage={account?.id === dialog.lastMessage?.author?.id}
          selected={dialogParam === dialog.id}
        />
      </li>
    )
  }

  return (
    <div className="dialogs">
      <div className="dialogs__header">
        <Search customStyles="dialogs__header-search-input" />
        <CreateDialog />
      </div>

      <ContentContainer loading={loading} errorMessage={errorMessage}>
        <ul className="dialogs__list list list--reset" onClick={selectDialog}>
          {dialogs?.length ? (
            dialogs.map(renderItem)
          ) : (
            <li className="dialogs__item dialogs__item--empty">Empty</li>
          )}
        </ul>
      </ContentContainer>
    </div>
  )
}
