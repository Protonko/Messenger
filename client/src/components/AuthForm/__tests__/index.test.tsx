import {shallow} from 'enzyme'
import {AuthForm, IPropsAuthForm} from 'components/AuthForm'

const shallowComponent = (props: IPropsAuthForm) => (
  shallow(<AuthForm {...props} />)
)

describe('AuthForm', () => {
  let props: IPropsAuthForm

  // beforeEach(() => {
  //   props = {}
  // })

  it('Match snapshot', () => {
    const component = shallowComponent(props)

    expect(component).toMatchSnapshot()
  })
}
)
