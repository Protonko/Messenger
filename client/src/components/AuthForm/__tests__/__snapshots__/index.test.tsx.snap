// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm Match snapshot 1`] = `
<form
  className="auth-form"
  onSubmit={[Function]}
>
  <Component
    type="h3"
  />
  <div
    className="auth-form__inputs"
  >
    <ul
      className="list list--reset auth-form__inputs-list"
    >
      <li
        className="auth-form__input"
        key="0"
      >
        <Input
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
          value=""
        />
      </li>
    </ul>
  </div>
  <div
    className="auth-form__buttons"
  >
    <ul
      className="list list--reset auth-form__buttons-list"
    >
      <li
        className="auth-form__buttons-item"
      >
        <Button
          additionalClassName="auth-form__button"
          text="Log In"
          type="submit"
        />
      </li>
      <li
        className="auth-form__buttons-item"
      >
        <Button
          additionalClassName="auth-form__button"
          modifier="linking"
          text="Create New Account"
        />
      </li>
    </ul>
  </div>
</form>
`;
